### ISSUE TITLE
‘relayWidgetMessages’ not in scope error after starting kernel

### ISSUE THREAD

**Issue by FelisNivalis** [IHaskell/IHaskell#1385]
I got the following error when I started a kernel
```
[I 00:16:53.865 NotebookApp] Creating new notebook in
[I 00:16:54.596 NotebookApp] Kernel started: 8818e04f-cf57-49f7-848b-23868661e6d8, name: haskell
[W 00:16:54.688 NotebookApp] 404 GET /static/components/codemirror/mode/ihaskell/ihaskell.js (127.0.0.1) 4.710000ms referer=http://localhost:8888/notebooks/Untitled.ipynb?kernel_name=haskell
[I 00:16:54.801 NotebookApp] Adapting from protocol version 5.0 (kernel 8818e04f-cf57-49f7-848b-23868661e6d8) to 5.3 (client).
ihaskell: Not in scope: ‘IHaskell.Eval.Widgets.relayWidgetMessages’
No module named ‘IHaskell.Eval.Widgets’ is imported.

[I 00:16:57.587 NotebookApp] KernelRestarter: restarting kernel (1/5), new random ports
WARNING:root:kernel 8818e04f-cf57-49f7-848b-23868661e6d8 restarted
[I 00:17:04.631 NotebookApp] Saving file at /Untitled.ipynb
```
but after it automatically restarted, the error disappeared and the kernel ran without problems any more.
so basically I still can use ihaskell, just the error is weird and a bit annoying. anyway it would be very nice if you could have a look.

versions:
- ihaskell: master branch
- ghcup: v0.1.18.0
- stack: 2.9.1
- python: 3.11.0
- system: wsl ubuntu jammy
and I installed ihaskell following the readme

**Comment by hugodro** [comment]
I'm hitting the same issue with kernel crashing due to IHaskell.Eval.Widget not being in scope, using ghc 9.4.5 and 9.4.7. I guess that's the first "extra" module that the ihaskell is trying to work with at runtime.

When Jupyter respawns a new ihaskell kernel, the error goes away (why??) but ihaskell can't find IHaskell modules anymore for an import from a notebook, nor extra packages like Data.Aeson.

Further more I noticed that builing ihaskell with ghc 9.2.5  works fine if I copy the generated binary in the .stack-work/.../Cabal-<version> folder by hand to ~/.local/bin, but it's also broken when the `stack install` does the job (taking the ihaskell binary from a build snapshot). If I dump symbols from the binary in the Cabal-<version> folder (ie the one that works fine) I can see the extra modules listed, but they don't show up anymore in the final built (ie the one that chokes trying to import IHaskell.Eval.Widgets). Maybe.that's a normal symbol strip performed by `stack install` before copy to ~/.local/bin but that's the only difference I can find between working/crashing ihaskell binaries.

I've ran Jupyter & ihaskell in debug mode, I don't see any difference between the contexts when the ihaskeld binary can import modulers from extra packages (ie built with ghc 9.2.5 and symbol-rich binary is hand-copied) vs it chokes (binary installed by `stack install`).

I've experimented on two separate machines (ubuntu 22.04), and I'm using ghcup to keep stack/cabal/... versions as currently recommended, and I've also tried building with system-ghc vs stack managed (x86_64-linux vs x86_64-linux-tinfo6) with no difference in results.

The ghc_package_path passed to the ihaskell at launch is correct (stack exec env) but I don't know if ihaskell is actually using that.

I'm really lost as to what is causing this problem. Is there anyone that can help?

**Comment by hugodro** [comment]
I've gotten rid of the problem by getting rid of all Haskell tool chain & Jupyter components I had install so far (ghcup, stack, jupyter-lab, ...), then reinstalled everything from scratch, started from a fresh git clone of the ihaskell project, then re-added my own code, and built using ghc 9.4.5 (ghcup installed). So the issue is somewhere in the build setup or jupyter runtime, and I really wish I knew where!

**Comment by vaibhavsagar** [comment]
I haven't experienced this with Nix, can I get you to try using that and see what happens? Examples are available at https://github.com/vaibhavsagar/notebooks.

**Comment by vaibhavsagar** [comment]
I suspect something in your build environment is causing this, which would make debugging difficult since I have no visibility into what's going on there. It would also explain why nuking everything and starting over seemed to fix it.

**Comment by hugodro** [comment]
Thanks for the quick follow-up!

Some additional info from today's debugging: first I was wrong, the full (?) cleaning up of Haskell toolchain & Jupyter packages didn't solve the problem, I was actually testing the already working setup when I thought the clean-up had worked out. 

I prefer to understand the issue rather than move to a Nix setup, which will just add more complexity in the overall setup. I use Haskell for many projects and I feel uneasy to know that sometimes it works, sometimes it doesn't and I can't understand why.  Further more I hope that a better understanding of how GHC.* packages handles the 'import' commands will lead to finding a way to add new extra packages on the fly rather than always having to having to "compile their presence" in the ihaskell build phase, which is a pain in the lower back.

I did some more tests today, copying the working binary produced on the "good" environment (a new user, with everything installed from scratch) to the "bad" environment (which somehow still has some broken setup).  And it leads me to think the problem is with the runtime config, as the working ihaskell binary on the good env shows the same kind of problem on the bad env as the one built in that "bad" env.

On the good side of things I did get the ihaskell in the good env to run as it used to in 9.2.5, and adding Hasql and using embedded sql within a notebook.  That's so far is my acid test #1, with acid test #2 being getting inline-js and run some js code in node.js as a co-execution support.  Also compiled with 9.4.7 without any problem.

I also figured out why the ihaskell kernel runs on the second launch in the bad environment: the Jupyter code sends a `comm_id: jupyter.widget.contol` at the first launch (to find out which ones are supported?), and that crashes ihaskell when it doesn't know how to resolve IHaskell.Eval.Widgets; but the 2nd kernel launch doesn't do that so the ihaskell doesn't choke before it's fully up.  It still can't resolve modules from any extra package, but at least  it sends an error back to jupyter instead of dying.

So anyways I'm pretty sure the problem is related to some path that GHC.* uses to get the packages info at runtime.  But I see nothing different between the 2 environments.

Also seen today: on the "good" env the `stack exec env` returns whatever is found in .stack/global-project/stack.yaml when the `stack exec jupyter ...` command isn't invoked from within the IHaskell project, which is expected behavior.  But I had compiled with resolver lts-21.6  but the global stack.yaml had lts-21.17, and still everything went fine; I had expected to see the ihaskell break on interface files (.hi). I confirmed that the kernel args had the 9.4.7 paths.  The same issue in the "bad" env creates _mismatched interface file versions (wanted "9045", got "9047")_, which I thought was normal.

That last event made me think the package databases aren't resolved the same way between the two env... Do you know how the paths could be shown by ihaskell when it starts? Alternatively if you let me know where in ihaskell I could catch the moment the jupyter.widgets.control is processed, I can try to do some debugging there.

**Comment by vaibhavsagar** [comment]
A couple of things in no particular order:

> EDIT: I've just done what I should have done much earlier and configured ihaskell to start with --env-file

I would strongly recommend against using `--env-file`, it's IMO a misfeature that was originally introduced to reduce Docker image size (https://github.com/IHaskell/IHaskell/pull/1252). However, the fact that it works supports my theory that the issue is somewhere within your environment.

> Further more I hope that a better understanding of how GHC.* packages handles the 'import' commands will lead to finding a way to add new extra packages on the fly rather than always having to having to "compile their presence" in the ihaskell build phase, which is a pain in the lower back.

IHaskell should already support loading extra packages without recompiling, what I used to do was install IHaskell globally with `ihaskell install --stack`, and then run `jupyter notebook` inside another project in a different directory (using the same Stack resolver) and it would have access to the project's dependencies (which you can confirm by running `:!ghc-pkg --list`). I'm also unclear on what you mean by "GHC.*", but more on that below

> I prefer to understand the issue rather than move to a Nix setup, which will just add more complexity in the overall setup.

The reason I suggested using Nix is that it does a much better job of isolating the system state, which again would either confirm or deny my theory about it having something to do with your environment, since if it's a bug in IHaskell then I would hope that our CI would reflect this and that's not what I'm seeing.

> So anyways I'm pretty sure the problem is related to some path that GHC.* uses to get the packages info at runtime. But I see nothing different between the 2 environments.

What you mean by "GHC.*"? I believe we've reimplemented most of how that stuff works in IHaskell, so whether or not you get an error from IHaskell is not a reflection on how the rest of the Haskell ecosystem or GHC itself works.

> Also seen today: on the "good" env the stack exec env returns whatever is found in .stack/global-project/stack.yaml when the stack exec jupyter ... command isn't invoked from within the IHaskell project, which is expected behavior. But I had compiled with resolver lts-21.6 but the global stack.yaml had lts-21.17, and still everything went fine; I had expected to see the ihaskell break on interface files (.hi). I confirmed that the kernel args had the 9.4.7 paths. The same issue in the "bad" env creates mismatched interface file versions (wanted "9045", got "9047"), which I thought was normal.

This is covered in the README [here](https://github.com/IHaskell/IHaskell#the-kernel-keeps-dying-ihaskell--stack), and once again it is a Stack-specific issue. The problem is that lately Stackage resolvers sometimes change the GHC minor version without changing the major version, so if two successive resolvers use e.g. GHC 9.4.7 they will be interoperable, but if one uses GHC 9.4.6 and the other one uses 9.4.7 then IHaskell will break.

> That last event made me think the package databases aren't resolved the same way between the two env... Do you know how the paths could be shown by ihaskell when it starts? Alternatively if you let me know where in ihaskell I could catch the moment the jupyter.widgets.control is processed, I can try to do some debugging there.

[`flushWidgetMessages`](https://github.com/IHaskell/IHaskell/blob/902c5609601693b27814e38c9712bce7c27c151c/src/IHaskell/Eval/Evaluate.hs#L545) seems to be the source of your issue, and it is called [here](https://github.com/IHaskell/IHaskell/blob/902c5609601693b27814e38c9712bce7c27c151c/main/Main.hs#L206) and [here](https://github.com/IHaskell/IHaskell/blob/902c5609601693b27814e38c9712bce7c27c151c/src/IHaskell/Eval/Evaluate.hs#L506). I suspect the first occurrence is what you're running into.

**Comment by hugodro** [comment]
Progress!
You provided the savier command: :!ghc-pkg list. 8-)

Together with :!stack exec env, this shows the situation clearly: for some (still unknown) reason the "stack exec env" that's executed from within ihaskell in the "bad" env just won't figure out the packages path that should be used for resolution. It especially doesn't know about the ihaskell-0.10.4.0 one, so the kernel crashes as soon as it tries to use proper IHaskell features. And whatever the directory the jupyter app is launched from, it's the same situation, unlike in the "good env". And using with --env-file will only work if jupyter is launched in the project directory in the "bad" env, while it will work no matter the directory

So it's definitely not a bug of ihaskell, but solving the config situation by re-organising the entire dev setup to Nix isn't convenient as a solution for me. Similarly to the message just above this one in the project issue list, I think having to organise the global stack env to fit IHaskell is not cooperative; and it would be more advantageous to be able to specify (in .ihaskell?) additional config params to get rid of .stack/global-project/stack.yaml dependency. I don't know the average use of ihaskell/jupyter but that's definitely the way I'd like to use it. That way both extreme isolation approach as Nix & others (as mine, I have around 20 active haskell projects all sharing the same toolset & global package database) would work out.

Beside that you were right that only a mention of extra packages is needed in ihaskell.cabal to have them accessible in the kernel. But I still need to do a stack install by hand to get them built into the exec environment.

But what I'd really like to be able to do is to install packages from within the notebook itself, as it's possible to do in the classic ipython kernel. That goes to having a good mastery of the Haskell compiler/interpreter (that's what I was referring to when using "GHC.*"). It's not available in ghci, but I've seen some Haskell code doing just that in the past (ie doing dynamic library discovery and linking within a running program rather than at startup), so somehow there's something that can be done in that direction.

I also got the inline-js working and I've got some node.js/npm packages connected. Do you have some idea (or maybe it's already built in jupyter/haskell) on how to specify values in the environment that the notebook/ihaskell can use so eg secrets & stuff as npm paths can be provided outside a notebook cells?

In conclusion the -env-file is the easiest solution to the package resolution issue, but having more configuration capability in IHaskell would be a great improvement for flexible operations.

Thanks again for your follow-up on this issue!!

**Comment by vaibhavsagar** [comment]
> Similarly to the message just above this one in the project issue list, I think having to organise the global stack env to fit IHaskell is not cooperative; and it would be more advantageous to be able to specify (in .ihaskell?) additional config params to get rid of .stack/global-project/stack.yaml dependency

If you're referring to https://github.com/IHaskell/IHaskell/issues/1387, you'll see that the desired behaviour is in fact already how this works.

> Beside that you were right that only a mention of extra packages is needed in ihaskell.cabal to have them accessible in the kernel. But I still need to do a stack install by hand to get them built into the exec environment.

That's not what I meant, I was referring to a workflow like https://github.com/IHaskell/IHaskell/issues/792#issuecomment-349186483

> But what I'd really like to be able to do is to install packages from within the notebook itself, as it's possible to do in the classic ipython kernel. That goes to having a good mastery of the Haskell compiler/interpreter (that's what I was referring to when using "GHC.*"). It's not available in ghci, but I've seen some Haskell code doing just that in the past (ie doing dynamic library discovery and linking within a running program rather than at startup), so somehow there's something that can be done in that direction.

I can see how this might be convenient for you, but I don't think I'd ever implement something like this in IHaskell.

> So it's definitely not a bug of ihaskell, but solving the config situation by re-organising the entire dev setup to Nix isn't convenient as a solution for me.

Understood, but since we agree that this isn't an IHaskell issue I think I can close this now.
